internal/hcp/registry/hcl.go

```diff
@@ -144,20 +145,72 @@ func NewHCLRegistry(config *hcl2template.PackerConfig, ui sdkpacker.Ui) (*HCLReg
 		return nil, diags
 	}
 
-	for _, source := range build.Sources {
-		bucket.RegisterBuildForComponent(source.String())
+	registry := &HCLRegistry{
+		configuration:  config,
+		bucket:         bucket,
+		ui:             ui,
+		metadata:       &MetadataStore{},
+		fullBuildNames: make(fullBuildNames),
 	}
 
 	ui.Say(fmt.Sprintf("Tracking build on HCP Packer with fingerprint %q", bucket.Version.Fingerprint))
 
-	return &HCLRegistry{
-		configuration: config,
-		bucket:        bucket,
-		ui:            ui,
-		metadata:      &MetadataStore{},
-	}, nil
+	return registry, registry.registerAllComponents()
+}
+
+func (h *HCLRegistry) registerAllComponents() hcl.Diagnostics {
+	var diags hcl.Diagnostics
+
+	// we currently support only one build block but it will change in the near future
+	for _, build := range h.configuration.Builds {
+		for _, source := range build.Sources {
+			if !h.fullBuildNames.set(build.Name, source.String()) {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Build name conflicts",
+					Subject:  &build.HCL2Ref.DefRange,
+					Detail: "When using HCP registry, build names are used as identifier. " +
+						"In this case, there is 2 or more build called " + h.fullBuildNames.get(build.Name, source.String()),
+				})
+			}
+		}
+	}
+
+	if diags.HasErrors() {
+		return diags
+	}
+
+	for sourceName, buildNames := range h.fullBuildNames {
+		for buildName := range buildNames {
+			h.bucket.RegisterBuildForComponent(h.fullBuildNames.get(buildName, sourceName))
+		}
+	}
+	return diags
 }
 
 func (h *HCLRegistry) Metadata() Metadata {
 	return h.metadata
 }
+
+type fullBuildNames map[string]map[string]struct{}
```

Comment:
I'm not necessarily against the idea of introducing a structure for this, but it feels a bit verbous to get them, I wonder if having functions directly on `HCLRegistry` wouldn't be better for this purpose?